### PR TITLE
[HP-1539] Add AWS functionality to cirrus

### PR DESCRIPTION
**Pull request by MichaelLukowski**
### New Features
- Added new functionality for AWS s3 API calls. This is for getting a presigned url for uploading a file, downloading a file, and downloading a file from a requester pays bucket

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -185,6 +202,14 @@ cirrus_config.update(**settings)
 
 *Still uses Google libraries for auth*
 
+## AWS Specific Implentation Details
+
+### Method for communication with AWS's API(s)
+
+For AWS you must bring your own Boto3 client that you have configured.
+
+You can then setup the AWS service and your client will be passed as an argument to the AWS api.
```
Comment (Avantol13) [inline README.md:211]:
```suggestion
You can then setup the AWS service and your client will be passed as an argument to the AWS API.
```

File: gen3cirrus/aws/services.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
+"""
+Amazon service for interacting with APIs
+"""
+
+import backoff
+import boto3
+from botocore.exceptions import ClientError
+from gen3cirrus.config import config
```
Comment (Avantol13) [inline gen3cirrus/aws/services.py:8]: I'd prefer if we break up the imports by newlines for 3 sections:

- built in libs
- external deps
- internal imports

File: gen3cirrus/aws/services.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+logger = get_logger(__name__, log_level="info")
+
+
+class AwsService(object):
+    """
+    Generic Amazon servicing using Boto3
+    """
+
+    def __init__(self, client):
```
Comment (Avantol13) [inline gen3cirrus/aws/services.py:20]: maybe just so it's super clear we should call this `boto3_client`?

File: gen3cirrus/aws/services.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        self.client = client
+
+    def downloadPresignedURL(self, bucket, key, expiration):
```
Comment (Avantol13) [inline gen3cirrus/aws/services.py:23]: docstrings on all new functions please

File: gen3cirrus/aws/utils.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        )
+        return None
+
+    try:
+        response = s3_client.generate_presigned_url(
+            m, Params={"Bucket": bucket_name, "Key": object_name}, ExpiresIn=expires
+        )
+
+    except ClientError as e:
+        logger.info(e)
```
Comment (Avantol13) [inline gen3cirrus/aws/utils.py:31]: is this really info, shouldn't it be error?

File: pyproject.toml [hunk 6]
```diff
@@ -16,6 +16,7 @@ google-cloud-storage = "*"
 google-api-python-client = "*"
 google-auth = "*"
 google-auth-httplib2 = "*"
+boto3 = "^1.34.128"
```
Comment (Avantol13) [inline pyproject.toml:19]: I'd prefer we leave this really open as "*" until there's a real conflict that requires us to limit it. Because this is a library imported  by other services, I want to be very flexible on the requirements until something actually requires limiting and we can't easily upgrade

File: test/test_utils.py [hunk 7]
```diff
@@ -9,6 +9,12 @@
     get_valid_service_account_id_for_user,
 )
 
+# for testing aws
+import boto3
```
Comment (Avantol13) [inline test/test_utils.py:13]: try to merge these with the previous imports and keep the 3 standard sections, alphabetize within the sections

File: test/test_utils.py [hunk 8]
```diff
@@ -121,3 +127,27 @@ def test_get_string_to_sign_no_optional_params():
     )
 
     assert result == ("GET\n" "\n" "\n" "1388534400\n" "/bucket/objectname")
+
+
+def test_aws_get_presigned_url():
```
Comment (Avantol13) [inline test/test_utils.py:132]: docstrings on tests please

**Review by Avantol13 — CHANGES_REQUESTED** [review]
Looks pretty good so far. See comments for minor things. Make sure you update the version in the pyproject.toml too - and then send me the Fence PR using this branch and we'll wait to merge this until we can verify everything is working in updated Fence

File: gen3cirrus/aws/services.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        """
+        Wrapper function for generating a presingned url for downloading an object
```
Comment (Avantol13) [inline gen3cirrus/aws/services.py:27]:
```suggestion
        Wrapper function for generating a presigned URL for downloading an object
```

and other spots in this file too

File: gen3cirrus/aws/utils.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    params["Bucket"] = bucket_name
+    params["Key"] = object_name
+
+    for key in additional_info:
+        params[key] = additional_info[key]
+
+    s3_client = client
+
+    if method == "get":
+        m = "get_object"
```
Comment (Avantol13) [inline gen3cirrus/aws/utils.py:37]: I'd prefer not to have single character names outside of simple loops. perhaps use `method` or something similar

File: gen3cirrus/aws/manager.py [hunk 11]
```diff
@@ -0,0 +1,7 @@
+class AwsCloudManager:
```
Comment (Avantol13) [inline gen3cirrus/aws/manager.py:1]: Let's just remove this. I don't see a clear need going forward and we can always add it later

File: README.md [hunk 12]
```diff
@@ -24,6 +24,23 @@ using the library like above.
...
+
+client = boto3.client()
+
+aws = AwsService(client)
+
+object = "test.txt"
+bucket = "testBucket"
+expiration = 3600
+
+url = aws.requestorPaysDownloadPresignedURL( bucket, object, expiration)
```
Comment (Avantol13) [inline README.md:41]: these functions need to be `camel_case`

File: gen3cirrus/aws/services.py (hunk 3 above)
Comment (Avantol13) [reply gen3cirrus/aws/services.py:26]: bump

File: gen3cirrus/aws/services.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    """
+
+    def __init__(self, client):
+        self.client = client
+
+    def download_presigned_url(self, bucket, key, expiration, additonal_info={}):
+        """
+        Wrapper function for generating a presigned URL for downloading an object
+        """
+        return generatePresignedURL(
```
Comment (Avantol13) [inline gen3cirrus/aws/services.py:33]: `lower_case_camel` on Python functions

File: gen3cirrus/aws/utils.py [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+
+import boto3
+from botocore.exceptions import ClientError
+
+from cdislogging import get_logger
+
+logger = get_logger(__name__, log_level="info")
+
+
+def generatePresignedURL(
```
Comment (Avantol13) [inline gen3cirrus/aws/utils.py:12]: `lower_case_camel`

File: gen3cirrus/aws/utils.py [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+    client, method, bucket_name, object_name, expires, additional_info={}
```
Comment (Avantol13) [inline gen3cirrus/aws/utils.py:13]: It's not safe to default a mutable type (like a dict) as a default value in function definitions. If you do it like this, the same mutable object gets used *every time the function is called* without explicitly providing that argument, which is (in almost every case) not the behavior you really intend b/c any changes to the variable get *persisted* between function calls.

In this case, I don't think you actually modify it, but it's considered a dangerous practice.

Check this out if you want to see this in action, the result is fun:

```python
def add_item_to_dict(item, my_dict={}):
    """Function to add an item to a dictionary."""
    if item in my_dict:
        my_dict[item] += 1
    else:
        my_dict[item] = 1
    return my_dict

# Call the function multiple times
print("First call:", add_item_to_dict('apple'))
print("Second call:", add_item_to_dict('banana'))
print("Third call:", add_item_to_dict('apple'))
```

The proper way to work around this pitfall is to use `None` and then set the mutable default in the actual function. `additional_info=None` in the definition, then in the function: `additional_info = additional_info or {}`

File: gen3cirrus/aws/utils.py [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+
+
+def generatePresignedURLRequesterPays(
+    client, bucket_name, object_name, expires, additional_info={}
+):
+    """
+    Function for generating a presigned URL only for requester pays buckets
+
+    Args:
+        client: s3 boto client
```
Comment (Avantol13) [inline gen3cirrus/aws/utils.py:102]: all these args need types. The format is `name (type): description`

File: test/test_utils.py [hunk 17]
```diff
@@ -1,5 +1,8 @@
+import boto3
+import botocore.session
 import pytest
 
+from botocore.stub import Stubber
```
Comment (Avantol13) [inline test/test_utils.py:5]: where is this used? I'm unsure how these tests are isolated from calling to AWS (and we do want to ensure that these are isolated and mocking AWS).

**Review by Avantol13 — CHANGES_REQUESTED** [review]

File: gen3cirrus/aws/utils.py (hunk 15 above)
Comment (mfshao) [reply gen3cirrus/aws/utils.py]: good suggestion! refactored

File: test/test_utils.py (hunk 17 above)
Comment (mfshao) [reply test/test_utils.py]: not sure where, so i just got rid of them 🤷‍♂️

File: test/test_utils.py [hunk 18]
```diff
@@ -121,3 +125,34 @@ def test_get_string_to_sign_no_optional_params():
...
 
     assert result == ("GET\n" "\n" "\n" "1388534400\n" "/bucket/objectname")
+
+
+def test_aws_get_presigned_url():
+    """
+    Test that we can get a presigned url from a bucket
+    """
+
+    s3 = boto3.client("s3", aws_access_key_id="", aws_secret_access_key="")
```
Comment (Avantol13) [inline test/test_utils.py:135]: I'm not seeing where this is mocked, so I'm actually not sure how this is passing

File: test/test_utils.py [hunk 19]
```diff
@@ -121,3 +125,34 @@ def test_get_string_to_sign_no_optional_params():
...
+    bucket = "test"
+    obj = "test-obj.txt"
+    expires = 3600
+
+    url = generate_presigned_url(s3, "get", bucket, obj, expires)
+
+    assert url != None
+
+
+def test_aws_get_presigned_url_requester_pays():
```
Comment (Avantol13) [inline test/test_utils.py:146]: we need more tests for the negative cases where boto throws an error

File: test/test_utils.py (hunk 18 above)
Comment (mfshao) [reply test/test_utils.py:135]: I'm not sure it need mocks 🤔 the `generate_presigned_url()` in boto3 is a local operation

**Review by Avantol13 — APPROVED** [review]

**Comment by mfshao** [comment]
@Avantol13 what do you think about https://github.com/uc-cdis/fence/pull/1173/files#r1727515677 ?

**Review by Avantol13 — APPROVED** [review]
